### DISCUSSION TITLE
Some source files don't include memdebug?

### DISCUSSION THREAD

**Discussion by jay** [curl/curl#6611] [General]
Most lib and src source files have memdebug as the last include, but some don't have it at all.

https://github.com/curl/curl/blob/2f33be817cbce6ad7a36f27dd7ada9219f13584c/src/tool_main.c#L53-L58

Isn't it always supposed to be the last include? Why is it in some files but not others? I will need it in src/tool_filetime.c for the Curl_safefree macro (#6602).

~~~
owner@ubuntu1604-x64-vm:~/curl$ grep -r --files-without-match --include='*.c' -F '#include "memdebug.h"' lib src | sort
lib/curl_ctype.c
lib/curl_des.c
lib/curl_endian.c
lib/curl_gethostname.c
lib/curl_multibyte.c
lib/curl_range.c
lib/easygetopt.c
lib/easyoptions.c
lib/inet_ntop.c
lib/inet_pton.c
lib/nonblock.c
lib/nwos.c
lib/parsedate.c
lib/progress.c
lib/select.c
lib/speedcheck.c
lib/splay.c
lib/strcase.c
lib/strtok.c
lib/strtoofft.c
lib/timeval.c
lib/version.c
lib/vquic/vquic.c
lib/vtls/bearssl.c
lib/vtls/rustls.c
lib/warnless.c
src/tool_filetime.c
src/tool_progress.c
src/tool_strdup.c
src/tool_writeout_json.c
~~~

**Comment by bagder** [comment]
Test 1132 is meant to verify that files (in `lib/`) that use memory functions include `memdebug.h` and `curl_memory.h`. It does this with the `tests/mem-include-scan.pl` script. If any of the files you list actually use memory functions then we need to figure out why the script fails and fix it!

**Comment by danielgustafsson** [comment]
> On 16 Feb 2021, at 08:14, Daniel Stenberg <notifications@github.com> wrote:
> 
> 
> Test 1132 is meant to verify that files (in lib/) that use memory functions include memdebug.h and curl_memory.h. It does this with the tests/mem-include-scan.pl script. If any of the files you list actually use memory functions then we need to figure out why the script fails and fix it!


Looking at a few select of thes files in lib/ they don't qualify as they dont have allocations. The files in src/ aren't included as 1132 only checks lib/

**Reply by jay** [reply]
The script also only checks for free|alloc|strdup to determine whether memdebug is needed but it does a lot more than that.

**Reply by bagder** [reply]
I don't think we actually use the memdebug system on src/ code these days, do we?

**Reply by jay** [reply]
Curl_safefree is used a number of times and it's defined in memdebug. If the file is included then it is in use is it not?  It's in most src files. memdebug appears to override fopen, recv, etc but that would only happen in the files where it's included.

**Reply by bagder** [reply]
the src files using memory functions are clearly in `log/memdump` after a test too...

**Comment by jay** [comment]
So the answer is that it's not always supposed to be included and as @bagder noted there's a script [memanalyze.pl](https://github.com/curl/curl/blob/curl-7_75_0/tests/memanalyze.pl) that determines which files need it. The PR I was working on #6602 actually needed functions and function-like macros exempt from memdebug, so I have adjusted the PR to reflect that.
